src/components/App.js

```diff
@@ -41,6 +62,54 @@ class App extends React.Component {
     this.setState({ filters });
   };
 
+  createNewBook = () => {
+    this.setState({ bookFormContent: {} });
+    this.setState({ editing: false });
+    this.setState({ showEditForm: true });
+  };
+
+  loadBook = isbn => {
+    const books = this.state.books;
+    const index = books.findIndex(b => b.isbn === isbn);
+    const bookFormContent = books[index];
+    this.setState({ bookFormContent });
+    this.setState({ editing: true });
+    this.setState({ showEditForm: true });
+  };
+
+  editBook = (key, value) => {
+    const bookFormContent = { ...this.state.bookFormContent };
+    if (value) {
+      bookFormContent[key] = value;
+    } else {
+      delete bookFormContent[key];
+    }
+    this.setState({ bookFormContent });
+  };
+
+  saveBook = () => {
+    const books = this.state.books;
+    const newBook = { ...this.state.bookFormContent };
+    const index = books.findIndex(b => b.isbn === newBook.isbn);
+    if (index >= 0) {
+      books.splice(index, 1);
+      books.push(newBook);
+    } else {
+      books.push(newBook);
+    }
```

Comment:
Ohhhhhhh!!! You're deleting the book from the old position and pushing it onto the back of the array! You aren't sorting by isbn. 

Okay, sorry, ignore what I said above.

This is fine then. You could still do the splice thing to replace the item (though change the 0 to 1) instead of pushing, but it's fine as is.

Reply:
Oh, duh, good catch I can use splice to inject the new book into the array in the same command.